File: MAF.m
function [results] = MAF(log,mafxaxis,mafyaxis,maftables,combmodes,logvars)
% [filestock,pathstock]=uigetfile('*.csv','Select Stock MAF_STD Table CSV File');
% current=csvread(fullfile(pathstock,filestock))

prompt = {'Confidence required to make change:'};
dlgtitle = 'MAF Correction Inputs';
dims = [1 50];
definput = {'0.95'};
answer = inputdlg(prompt,dlgtitle,dims,definput)
conf=str2num(answer{1});
maxcount=100


%% Create Derived Values

log.MAP=log.MAP.*10

if ~any(contains(logvars,'LAM_DIF'))
    log.LAM_DIF=1./log.LAMBDA_SP-1./log.LAMBDA
end

if any(contains(logvars,'FAC_LAM_OUT'))
    log.ADD_MAF=log.FAC_LAM_OUT-log.LAM_DIF 
else
    log.ADD_MAF=log.STFT+log.LTFT-log.LAM_DIF 
end

if any(contains(logvars,'MAF_COR'))
    log.ADD_MAF=log.ADD_MAF+log.MAF_COR
end

% log(log.EngState<2,:) = [];
% log(log.EngState>4,:) = [];
log(log.state_lam~=1,:) = [];
% log(log.LAMBDA==0,:) = [];


%% Read Axis Values
% stock=csvread(fullfile(getcurrentdir,"VE_stock.csv"))
% mafxaxis=current(1,2:width(current))
% mafyaxis=reshape(current(2:height(current),1),1,[])
xlabels=string(mafxaxis)
ylabels=string(mafyaxis)
% current(1,:)=[]
% current(:,1)=[]
%% Create Bins

xedges(1)=0;
xedges(length(mafxaxis)+1)=inf;
for i=1:length(mafxaxis)-1;
    xedges(i+1)=(mafxaxis(i)+mafxaxis(i+1))/2;
end

yedges(1)=0;
yedges(length(mafyaxis)+1)=inf;
for i=1:length(mafyaxis)-1;
    yedges(i+1)=(mafyaxis(i)+mafyaxis(i+1))/2;
end



%% Discretize MAF

X=discretize(log.RPM,xedges);
Y=discretize(log.MAP,yedges);
log.X=X;
log.Y=Y;

%% Initialize matrixes

for IDX=0:3

    SUM=zeros(length(mafyaxis),length(mafxaxis));
    COUNT=SUM;
    AVG=SUM;
    columns= zeros(12,8)
    rows=columns
    current=maftables{IDX+1}
    IDXmodes=find(combmodes==IDX)-1

    if any(contains(logvars,'MAF_COR'))
        test=current
    else
        test=zeros(length(mafyaxis),length(mafxaxis))
    end

    temp1=log
    diffcmb=setdiff(log.CMB,IDXmodes)
    for k=1:length(diffcmb)
        temp1(temp1.CMB==diffcmb(k),:)=[]
    end

    for i=1:length(mafxaxis)
        temp=temp1;
        temp(temp.X~=i,:)=[];
        tempcol = lsq_lut_piecewise( temp.MAP, temp.ADD_MAF, mafyaxis )
        if size(tempcol)==[8,1]
            columns(:,i) = tempcol
        end
    end
    
    for j=1:length(mafyaxis)
        temp=temp1;
        temp(temp.Y~=j,:)=[];
        temprow = lsq_lut_piecewise( temp.RPM, temp.ADD_MAF, mafxaxis )
        if size(temprow)==[8,1]
            rows(j,:) = temprow
        end
    end
    
    blend=(columns+rows)/2
    interpfac=.25
    
    
    for i=1:length(mafxaxis)
        for j=1:length(mafyaxis)
            temp=temp1;
            temp(temp.X~=i,:)=[];
            temp(temp.Y~=j,:)=[];
            AVGtemp=mean(temp.ADD_MAF)
            COUNT(j,i)=height(temp)
            if COUNT(j,i)>3
                ci=paramci(fitdist(temp.ADD_MAF,'Normal'),'Alpha',conf)
                sigma(j,i)=ci(2,2)
                low(j,i)=ci(1,1)
                high(j,i)=ci(2,1)
                if low(j,i)>test(j,i)
                    AVG(j,i)=(blend(j,i)*interpfac+low(j,i)*(1-interpfac))
                elseif high(j,i)<test(j,i)
                    AVG(j,i)=(blend(j,i)*interpfac+high(j,i)*(1-interpfac))
                    else
                    AVG(j,i)=test(j,i)
                end
            else
                AVG(j,i)=test(j,i)
            end
        end
    end
    
    COUNT(COUNT>maxcount)=[maxcount]
    COUNT=COUNT/maxcount
    
    COUNT(isnan(COUNT))=0
    AVG(isnan(AVG))=0
    CHANGE=(AVG-test).*COUNT
    NEW=round((current+CHANGE)*5.12,0)/5.12
    
    my_field = strcat('IDX',num2str(IDX))
    results.(my_field)=array2table(NEW,'VariableNames',xlabels,'RowNames',ylabels)
    fig3 = uifigure('Name',strcat('MAF Results IDX []: ',num2str(IDX)),"Position",[500 500 760 360]);
    uit = uitable(fig3, "Data",results.(my_field), "Position",[20 20 720 320]);
    styleIndices = (CHANGE);
    [row,col] = find(styleIndices);
    s = uistyle("BackgroundColor",'gr');
    addStyle(uit,s,"cell",[row,col]);
end

end

File: MAF.py
# C:/Users/Sam/PycharmProjects/YAKtunerCONVERTED/MAF.py

"""
Mass Airflow (MAF) Correction Tuning Module

This module contains pure, non-UI functions to analyze engine logs and recommend
adjustments to the four primary MAF correction tables (IDX0-IDX3).
"""

import numpy as np
import pandas as pd
from scipy import stats, interpolate


# --- Helper Functions ---

def _process_and_filter_maf_data(log, logvars):
    """
    A pure function to prepare and filter log data for MAF tuning using the new
    unified correction formula. This now includes step-wise filtering with specific warnings.
    """
    warnings = []
    df = log.copy()

    # --- Step-wise filtering with specific warnings ---
    # Step 1: Oil Temperature Filter
    if "OILTEMP" in logvars:
        df = df[df['OILTEMP'] > 180].copy()
        if df.empty:
            warnings.append("No data remained after filtering for Oil Temperature > 180°F.")
            return pd.DataFrame(), warnings

    # Step 2: Unit conversion for MAP (kPa to hPa/mbar)
    if 'MAP' in df.columns:
        df.loc[:, 'MAP'] = df['MAP'] * 10
    else:
        warnings.append("Log variable 'MAP' not found. MAF analysis will likely fail.")
        return pd.DataFrame(), warnings

    # Step 3: Pre-emptive check for division by zero
    if 'LAMBDA_SP' in df.columns:
        zero_sp_mask = df['LAMBDA_SP'] == 0
        if zero_sp_mask.any():
            warnings.append("Log contains rows where 'LAMBDA_SP' is zero. These rows are being ignored.")
            df = df[~zero_sp_mask].copy()
            if df.empty:
                warnings.append("No data remained after filtering for zero-value 'LAMBDA_SP'.")
                return pd.DataFrame(), warnings
    else:
        warnings.append("Log variable 'LAMBDA_SP' not found. Cannot calculate fuel error.")
        return pd.DataFrame(), warnings

    # --- New Unified Correction Formula ---
    required_vars = ['LAMBDA', 'LAMBDA_SP']
    if not all(v in df.columns for v in required_vars):
        raise ValueError(f"MAF analysis requires essential log variables: {required_vars}")

    # Get all potential correction factors, with defaults
    maf_cor = df.get('MAF_COR', 0.0)
    if 'MAF_COR' not in logvars: warnings.append("Log 'MAF_COR' for best accuracy.")

    mff_cor = df.get('MFF_COR', 1.0)
    if 'MFF_COR' not in logvars: warnings.append("Log 'MFF_COR' for best accuracy.")

    # --- FIX: Simplified retrieval of STFT and LTFT variables ---
    fac_lam_out = df.get('FAC_LAM_OUT', 0.0)
    stft = df.get('STFT', 0.0)
    fac_mff_add = df.get('FAC_MFF_ADD', 0.0)
    ltft = df.get('LTFT', 0.0)
    # --- END FIX ---

    # --- FIX: Simplified degradation logic for LTFT ---
    ltft_correction_term = 1.0
    if 'FAC_MFF_ADD' in logvars:
        ltft_correction_term = (1 + fac_mff_add / 100)
        warnings.append("Using 'FAC_MFF_ADD' for LTFT correction.")
    elif 'LTFT' in logvars:
        ltft_correction_term = (1 + ltft / 100)
        warnings.append("Using 'LTFT' as fallback for LTFT correction.")
    else:
        warnings.append("No suitable LTFT correction variable found. Assuming neutral LTFT correction (1.0).")
    # --- END FIX ---

    # --- FIX: Simplified degradation logic for STFT ---
    stft_correction_term = 1.0
    if 'FAC_LAM_OUT' in logvars:
        stft_correction_term = (1 + fac_lam_out / 100)
        warnings.append("Using 'FAC_LAM_OUT' for STFT correction.")
    elif 'STFT' in logvars:
        stft_correction_term = (1 + stft / 100)
        warnings.append("Using 'STFT' as fallback for STFT correction.")
    else:
        warnings.append("No suitable STFT correction variable found. Assuming neutral STFT correction (1.0).")
    # --- END FIX ---

    total_ecu_factor = (1 + maf_cor / 100) * stft_correction_term * mff_cor * ltft_correction_term
    measured_error = df['LAMBDA'] / df['LAMBDA_SP']
    target_factor = total_ecu_factor * measured_error

    maf_cor_new = target_factor - 1
    df.loc[:, 'ADD_MAF'] = maf_cor_new

    return df, warnings


def _create_bins(log, mafxaxis, mafyaxis):
    """Discretizes log data into bins based on MAF map axes."""
    xedges = [0] + [(mafxaxis[i] + mafxaxis[i + 1]) / 2 for i in range(len(mafxaxis) - 1)] + [np.inf]
    yedges = [0] + [(mafyaxis[i] + mafyaxis[i + 1]) / 2 for i in range(len(mafyaxis) - 1)] + [np.inf]

    log.loc[:, 'X'] = pd.cut(log['RPM'], bins=xedges, labels=False, duplicates='drop')
    log.loc[:, 'Y'] = pd.cut(log['MAP'], bins=yedges, labels=False, duplicates='drop')
    return log


def _fit_surface_maf(log_data, mafxaxis, mafyaxis):
    """
    Fits a surface to the MAF correction data, gracefully handling low-dimensional and
    non-finite data to prevent Qhull and other interpolation errors.
    """
    if log_data.empty:
        return np.zeros((len(mafyaxis), len(mafxaxis)))

    # Proactively clean data to remove rows with non-finite values (NaN or inf)
    finite_mask = np.isfinite(log_data['RPM']) & np.isfinite(log_data['MAP']) & np.isfinite(log_data['ADD_MAF'])
    clean_log_data = log_data[finite_mask]

    if clean_log_data.empty or len(clean_log_data) < 3:
        return np.zeros((len(mafyaxis), len(mafxaxis)))

    points = clean_log_data[['RPM', 'MAP']].values
    values = clean_log_data['ADD_MAF'].values
    grid_x, grid_y = np.meshgrid(mafxaxis, mafyaxis)

    # Check for variation in both dimensions
    x_variation = np.ptp(points[:, 0]) > 1e-6
    y_variation = np.ptp(points[:, 1]) > 1e-6

    if x_variation and y_variation:
        try:
            fitted_surface = interpolate.griddata(points, values, (grid_x, grid_y), method='linear')
        except Exception:
            fitted_surface = interpolate.griddata(points, values, (grid_x, grid_y), method='nearest')
    elif x_variation:
        print("Warning: MAF log data is flat in the Y-dimension (MAP). Using 1D interpolation.")
        unique_x, mean_values_idx = np.unique(points[:, 0], return_inverse=True)
        avg_values = np.bincount(mean_values_idx, weights=values) / np.bincount(mean_values_idx)
        interp_values = np.interp(mafxaxis, unique_x, avg_values, left=avg_values[0], right=avg_values[-1])
        fitted_surface = np.tile(interp_values, (len(mafyaxis), 1))
    elif y_variation:
        print("Warning: MAF log data is flat in the X-dimension (RPM). Using 1D interpolation.")
        unique_y, mean_values_idx = np.unique(points[:, 1], return_inverse=True)
        avg_values = np.bincount(mean_values_idx, weights=values) / np.bincount(mean_values_idx)
        interp_values = np.interp(mafyaxis, unique_y, avg_values, left=avg_values[0], right=avg_values[-1])
        fitted_surface = np.tile(interp_values, (len(mafxaxis), 1)).T
    else:
        print("Warning: MAF log data has no variation in X or Y dimensions. Using mean value.")
        mean_value = np.mean(values)
        fitted_surface = np.full((len(mafyaxis), len(mafxaxis)), mean_value)

    # Fill any remaining NaNs from the griddata process
    nan_mask = np.isnan(fitted_surface)
    if np.any(nan_mask):
        nearest_fill = interpolate.griddata(points, values, (grid_x[nan_mask], grid_y[nan_mask]), method='nearest')
        if nearest_fill is not None and not np.all(np.isnan(nearest_fill)):
            fitted_surface[nan_mask] = nearest_fill

    return np.nan_to_num(fitted_surface)


def _calculate_maf_correction(log_data, blend_surface, old_table, mafxaxis, mafyaxis, confidence, additive_mode=False):
    """
    Applies confidence interval logic to determine the final correction table.
    """
    new_table = old_table.copy()
    max_count = 80
    interp_factor = 0.5

    for i in range(len(mafxaxis)):
        for j in range(len(mafyaxis)):
            cell_data = log_data[(log_data['X'] == i) & (log_data['Y'] == j)]

            # Proactively clean data for statistical analysis
            cell_data = cell_data[np.isfinite(cell_data['ADD_MAF'])]

            count = len(cell_data)

            if count > 3:
                mean, std_dev = stats.norm.fit(cell_data['ADD_MAF'])
                surface_val = blend_surface[j, i]
                target_val = (surface_val * interp_factor) + (mean * (1 - interp_factor))
                low_ci, high_ci = stats.norm.interval(confidence, loc=target_val,
                                                      scale=std_dev if std_dev > 0 else 1e-9)
                current_val_from_table = old_table[j, i]
                comparison_val = 0.0 if additive_mode else current_val_from_table

                if not (low_ci <= comparison_val <= high_ci):
                    weight = min(count, max_count) / max_count
                    change_amount = (target_val - comparison_val) * weight
                    new_table[j, i] = current_val_from_table + change_amount

    recommended_table = np.round(new_table * 5.12) / 5.12
    return recommended_table


# --- Main Orchestrator Function ---
def run_maf_analysis(log, mafxaxis, mafyaxis, maftables, combmodes_MAF, logvars):
    """
    Main orchestrator for the MAF tuning process. A pure computational function.
    """
    try:
        print(" -> Initializing MAF analysis...")
        params = {'confidence': 0.6}

        print(" -> Preparing MAF data from logs...")
        processed_log, warnings = _process_and_filter_maf_data(log, logvars)

        additive_mode = 'MAF_COR' not in logvars
        if additive_mode:
            warnings.append("MAF_COR not found in logs. Switching to additive correction mode.")

        if processed_log.empty:
            # The warnings list will already contain the specific reason from the filter function
            return {'status': 'Failure', 'warnings': warnings, 'results_maf': None}

        print(" -> Creating data bins from MAF axes...")
        log_binned = _create_bins(processed_log, mafxaxis, mafyaxis)

        if 'CMB' not in log_binned.columns:
            warnings.append("Log variable 'CMB' (Combination Mode) not found. Cannot determine which MAF table to correct.")
            return {'status': 'Failure', 'warnings': warnings, 'results_maf': None}

        results = {}
        for idx in range(4):
            print(f" -> Processing MAF Table IDX{idx}...")
            current_table = maftables[idx]
            idx_modes = np.where(combmodes_MAF == idx)[0]
            log_filtered = log_binned[log_binned['CMB'].isin(idx_modes)].copy()

            log_filtered = log_filtered[
                np.isfinite(log_filtered['RPM']) & np.isfinite(log_filtered['MAP']) & np.isfinite(log_filtered['ADD_MAF'])]

            if log_filtered.empty:
                warnings.append(f"No valid data found for MAF Table IDX{idx}. This table will not be changed.")
                xlabels = [str(x) for x in mafxaxis]
                ylabels = [str(y) for y in mafyaxis]
                results[f'IDX{idx}'] = pd.DataFrame(current_table, columns=xlabels, index=ylabels)
                continue

            print(f"   -> Fitting 3D surface for IDX{idx}...")
            blend_surface = _fit_surface_maf(log_filtered, mafxaxis, mafyaxis)

            print(f"   -> Calculating correction map for IDX{idx}...")
            recommended_table = _calculate_maf_correction(
                log_filtered, blend_surface, current_table, mafxaxis, mafyaxis, params['confidence'],
                additive_mode=additive_mode
            )

            xlabels = [str(x) for x in mafxaxis]
            ylabels = [str(y) for y in mafyaxis]
            results[f'IDX{idx}'] = pd.DataFrame(recommended_table, columns=xlabels, index=ylabels)

        print(" -> MAF analysis complete.")
        return {
            'status': 'Success',
            'warnings': warnings,
            'results_maf': results
        }
    except Exception as e:
        return {
            'status': 'Failure',
            'warnings': [f"A critical error occurred inside the MAF module: {e}"],
            'results_maf': None
        }